### PR TITLE
WIP Feature: CODEOWNERS policy

### PR DESCRIPTION
**Pull request by Gby56**
Implements https://github.com/ossf/allstar/issues/284
I also had to bump `github.com/google/go-github` github REST client to v50, to access the function `GetCodeownersErrors` as it was not exposed in the API of the v43

### CODE REVIEW

**Comment by Gby56** [comment]
Hello @jeffmendoza :)
I ran this locally to test things out, seems like it's good, I'm just not sure about what tests to implement yet
It's pretty simple, the API returns a 404 if no CODEOWNERS is found. The GraphQL API was a nightmare to figure out, so I had to use the REST one, and bump it (as explained above)

**Comment by coheigea** [comment]
I'd recommend doing the version bump in a separate initial PR @Gby56 . Curious as to why dependabot didn't pick it up?

File: pkg/policies/codeowners/codeowners.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+	log.Info().
+		Str("org", owner).
+		Str("repo", repo).
+		Str("area", polName).
+		Bool("enabled", enabled).
+		Msg("Check repo enabled")
+
+	codeownererrors, _, err := rep.GetCodeownersErrors(context.Background(), owner, repo)
+
+	if err != nil {
```
Comment (jeffmendoza) [inline pkg/policies/codeowners/codeowners.go:122]: Probably want to differentiate here between an expected 404 and an unexpected error. See other code for examples: https://github.com/ossf/allstar/search?q=http.StatusNotFound&type=code

File: pkg/policies/codeowners/codeowners.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+			Enabled:    enabled,
+			Pass:       true,
+			NotifyText: "",
+		}, nil
+	}
+
+	return &policydef.Result{
+		Enabled:    enabled,
+		Pass:       false,
+		NotifyText: "CODEOWNERS file not present.\n" + notifyText,
```
Comment (jeffmendoza) [inline pkg/policies/codeowners/codeowners.go:141]: Wouldn't this case be that the file is present, but has errors?

File: pkg/policies/codeowners/codeowners.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+type mergedConfig struct {
+	Action string
+}
+
+type details struct {
+	Enabled bool
+	URL     string
+}
```
Comment (jeffmendoza) [inline pkg/policies/codeowners/codeowners.go:64]: The details struct is what you would pass back as a part of the policydef.Result. You probably want a bool if it was found or not (404 or not), and a count of errors if found.

**Review by jeffmendoza — COMMENTED** [review]
Looking good so far. I'll do another pass when the lint/test errors are fixed. Looks like some may be due to updating the github client verion.

**Comment by jeffmendoza** [comment]
Thanks for bumping the github client version, dependabot doesn't because they usually bump the major version, and it can cause incompatabilites. A separate PR for that alone would be ok.

**Comment by Gby56** [comment]
Thanks a lot ! I'll do a separate PR for the bump tomorrow then :) and yes you're right, I should count the codeowner errors and maybe still create an issue when it's broken, even if at least it exists (i think it's best)

File: pkg/enforce/enforce.go [hunk 4]
```diff
@@ -130,6 +130,16 @@ func EnforceAll(ctx context.Context, ghc ghclients.GhClientsInterface) (EnforceA
...
+			// For debug usage, you can filter repositories to test against a single one
+			// TODO: Implement a cli flag option to specific a repository, and/or a policy you want to run, useful for testing.
+			// var found github.Repository
+			// for _, r := range repos {
+			// 	if *r.Name == "daikon-ee" {
+			// 		found = *r
+			// 	}
+			// }
+			// repos = []*github.Repository{}
+			// repos = append(repos, &found)
```
Comment (coheigea) [inline pkg/enforce/enforce.go:142]: Implement as new PR please :-)

File: pkg/policies/codeowners/codeowners.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
+// Copyright 2021 Allstar Authors
```
Comment (coheigea) [inline pkg/policies/codeowners/codeowners.go:1]: 2023

File: pkg/policies/codeowners/codeowners_test.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
+// Copyright 2021 Allstar Authors
```
Comment (coheigea) [inline pkg/policies/codeowners/codeowners_test.go:1]: 2023

File: pkg/policies/codeowners/codeowners.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+	"github.com/rs/zerolog/log"
+)
+
+const configFile = "codeowners.yaml"
+const polName = "CODEOWNERS"
+
+const notifyText = `A CODEOWNERS file can give users information about who is responsible for the maintenance of the repository, or specific folders/files. This is different the access control/permissions on a repository.
+
+To fix this, add a CODEOWNERS file to your repository, following the official Github documentation and maybe your company's policy.
+https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners`
```
Comment (coheigea) [inline pkg/policies/codeowners/codeowners.go:34]: Put the backtick on a new line

File: pkg/policies/codeowners/codeowners.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+// OrgConfig is the org-level config definition for Branch Protection.
+type OrgConfig struct {
+	// OptConfig is the standard org-level opt in/out config, RepoOverride applies to all
+	// BP config.
+	OptConfig config.OrgOptConfig `json:"optConfig"`
+
+	// Action defines which action to take, default log, other: issue...
+	Action string `json:"action"`
+
+	//TODO add default contents for "fix" action
```
Comment (coheigea) [inline pkg/policies/codeowners/codeowners.go:45]: We should allow an exemption on a per-repo basis, so when specified we don't require Codeowners for the given repo. But this could be a separate PR if you want.

**Review by coheigea — CHANGES_REQUESTED** [review]
